receiver/awsfirehosereceiver/unmarshaler/cwmetricstream/metricsbuilder.go

```diff
@@ -0,0 +1,157 @@
+		rmb.metricBuilders[metricKey] = mb
+	}
+	mb.AddDataPoint(metric)
+}
+
+// Build creates the pdata.ResourceMetrics from the added metrics.
+func (rmb *resourceMetricsBuilder) Build() pdata.ResourceMetrics {
+	rm := pdata.NewResourceMetrics()
+	ilm := rm.InstrumentationLibraryMetrics().AppendEmpty()
+	rmb.toResource().CopyTo(rm.Resource())
+	for _, mb := range rmb.metricBuilders {
+		mb.Build().CopyTo(ilm.Metrics().AppendEmpty())
+	}
+	return rm
+}
+
+// toResource creates a pdata.Resource from the fields in the resourceMetricsBuilder.
+// Splits the namespace into service.namespace/service.name if prepended by AWS/.
+func (rmb *resourceMetricsBuilder) toResource() pdata.Resource {
+	resource := pdata.NewResource()
+	attributes := resource.Attributes()
+	attributes.InsertString(conventions.AttributeCloudProvider, conventions.AttributeCloudProviderAWS)
+	attributes.InsertString(conventions.AttributeCloudAccountID, rmb.accountID)
+	attributes.InsertString(conventions.AttributeCloudRegion, rmb.region)
+	splitNamespace := strings.SplitN(rmb.namespace, namespaceDelimiter, 2)
+	if len(splitNamespace) == 2 && strings.EqualFold(splitNamespace[0], conventions.AttributeCloudProviderAWS) {
+		attributes.InsertString(conventions.AttributeServiceNamespace, splitNamespace[0])
+		attributes.InsertString(conventions.AttributeServiceName, splitNamespace[1])
+	} else {
+		attributes.InsertString(conventions.AttributeServiceName, rmb.namespace)
+	}
+	attributes.InsertString(attributeAWSCloudWatchMetricStreamName, rmb.metricStreamName)
+	return resource
+}
+
+// toMetricKey creates a key based on the metric name and dimensions to
+// keep metrics with different dimensions separate.
+func (rmb *resourceMetricsBuilder) toMetricKey(metric cWMetric) string {
+	return fmt.Sprintf("%s::%v", metric.MetricName, metric.Dimensions)
+}
+
+type metricBuilder struct {
+	name       string
+	unit       string
+	dataPoints pdata.HistogramDataPointSlice
+	timestamps map[int64]bool
+}
+
+func newMetricBuilder(name, unit string) *metricBuilder {
+	return &metricBuilder{
+		name:       name,
+		unit:       unit,
+		dataPoints: pdata.NewHistogramDataPointSlice(),
+		timestamps: make(map[int64]bool),
+	}
+}
+
+// AddDataPoint adds the metric as a datapoint if a metric for that timestamp
+// hasn't already been added
+func (mb *metricBuilder) AddDataPoint(metric cWMetric) {
+	if _, ok := mb.timestamps[metric.Timestamp]; !ok {
+		mb.toDataPoint(metric).CopyTo(mb.dataPoints.AppendEmpty())
+		mb.timestamps[metric.Timestamp] = true
+	}
+}
+
+// Build builds the pdata.Metric with the data points that were added
+// with AddDataPoint
+func (mb *metricBuilder) Build() pdata.Metric {
+	metric := pdata.NewMetric()
+	metric.SetName(mb.name)
+	metric.SetUnit(mb.unit)
+	metric.SetDataType(pdata.MetricDataTypeHistogram)
+	metric.Histogram().SetAggregationTemporality(pdata.MetricAggregationTemporalityDelta)
+	mb.dataPoints.CopyTo(metric.Histogram().DataPoints())
```

Comment:
```suggestion
	mb.dataPoints.MoveAndAppendTo(metric.Histogram().DataPoints())
```

This will move the data point slice into the new metric without copying/allocation.